frame/bridge/ethereum/issuing/src/lib.rs

```diff
@@ -21,275 +21,351 @@
+use dp_contract::{
+	ethereum_backing::{EthereumBacking, EthereumLockEvent, EthereumRegisterEvent},
+	mapping_token_factory::{
+		MappingTokenFactory as mtf, TokenBurnInfo, TokenRegisterInfo, BURN_ACTION, REGISTER_ACTION,
+	},
+};
 use ethereum_primitives::{
 	receipt::{EthereumTransactionIndex, LogEntry},
-	EthereumAddress,
+	EthereumAddress, U256,
 };
-use types::*;
-
-pub trait Config: dvm_ethereum::Config {
-	type Event: From<Event<Self>> + Into<<Self as frame_system::Config>::Event>;
-
-	type PalletId: Get<PalletId>;
-
-	type RingCurrency: LockableCurrency<Self::AccountId, Moment = Self::BlockNumber>;
-
-	type EthereumRelay: EthereumReceipt<Self::AccountId, RingBalance<Self>>;
-	type EcdsaAuthorities: RelayAuthorityProtocol<Self::BlockNumber, Signer = EthereumAddress>;
-
-	type WeightInfo: WeightInfo;
-}
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	mod types {
+		use crate::*;
+
+		pub type AccountId<T> = <T as frame_system::Config>::AccountId;
+		pub type EthereumReceiptProofThing<T> = <<T as Config>::EthereumRelay as EthereumReceipt<
+			AccountId<T>,
+			RingBalance<T>,
+		>>::EthereumReceiptProofThing;
+		pub type RingBalance<T> = <<T as Config>::RingCurrency as Currency<AccountId<T>>>::Balance;
+	}
+	pub use types::*;
+
+	#[pallet::config]
+	#[pallet::disable_frame_system_supertrait_check]
+	pub trait Config: dvm_ethereum::Config {
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+		type RingCurrency: LockableCurrency<Self::AccountId, Moment = Self::BlockNumber>;
+		type EthereumRelay: EthereumReceipt<Self::AccountId, RingBalance<Self>>;
+		type EcdsaAuthorities: RelayAuthorityProtocol<Self::BlockNumber, Signer = EthereumAddress>;
+		type WeightInfo: WeightInfo;
+	}
 
-decl_error! {
+	#[pallet::error]
 	/// Issuing pallet errors.
-	pub enum Error for Module<T: Config> {
+	pub enum Error<T> {
 		/// Invalid Issuing System Account
 		InvalidIssuingAccount,
 		/// assert ar
 		AssetAR,
-		/// LogEntryNE
-		LogEntryNE,
-		/// EthLogPF
-		EthLogPF,
 		/// StringCF
 		StringCF,
-		/// Unit
-		UintCF,
-		/// Address - CONVERSION FAILED
-		AddressCF,
 		/// encode erc20 tx failed
 		InvalidEncodeERC20,
-		/// encode mint tx failed
-		InvalidMintEncoding,
 		/// invalid ethereum address length
 		InvalidAddressLen,
 		/// decode input value error
 		InvalidInputData,
+		/// decode ethereum event failed
+		DecodeEventFailed,
+		/// invalid input length
+		InvalidInput,
 	}
-}
 
-decl_event! {
-	pub enum Event<T>
-	where
-		AccountId = AccountId<T>,
-	{
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	#[pallet::metadata(T::AccountId = "AccountId")]
+	pub enum Event<T: Config> {
 		/// register new erc20 token
-		RegisterErc20(AccountId, EthereumAddress, EthereumTransactionIndex),
+		RegisterErc20(EthereumAddress, EthereumTransactionIndex),
```

Comment:
Can we use `TokenRegistered(u8, EthereumAddress, EthereumAddress, EthereumAddress)` to replace `RegisterErc20`, I see there is a type field in the `TokenRegistered`  event.

Reply:
No，`RegisterErc20` and `TokenRegistered` are two different events. `TokenRegistered`  is deposited when the token registration is finished.

Reply:
`RegisterErc20` says the first step of registering token, `TokenRegistered` says the registration process successfully?

Reply:
Yes

Reply:
Good name.